libs/nio/src/main/java/org/elasticsearch/nio/NioSelector.java

```diff
@@ -284,20 +271,29 @@ public void scheduleForRegistration(NioChannel channel) {
     }
 
     /**
-     * Queues a write operation directly in a channel's buffer. Channel buffers are only safe to be accessed
-     * by the selector thread. As a result, this method should only be called by the selector thread.
+     * Queues a write operation directly in a channel's buffer. If this channel does not have pending writes
+     * already, the channel will be flushed. Channel buffers are only safe to be accessed by the selector
+     * thread. As a result, this method should only be called by the selector thread. If this channel does
+     * not have pending writes already, the channel will be flushed.
      *
      * @param writeOperation to be queued in a channel's buffer
      */
-    public void queueWriteInChannelBuffer(WriteOperation writeOperation) {
+    public void writeToChannel(WriteOperation writeOperation) {
         assertOnSelectorThread();
         SocketChannelContext context = writeOperation.getChannel();
+        boolean shouldFlush = context.readyForFlush() == false;
```

Comment:
this logic doesn't look intuitive to me. why should we flush if the context is not ready for it? maybe it's a naming issue?